### PR TITLE
perf: vectorize finished-beam cache in beam search

### PR DESCRIPTION
**Pull request by sword-nan**
## Description

This PR speeds up beam search in Casanovo by keeping the cache of finished beams on the GPU and replacing per-step Python `heapq` operations with vectorized tensor ops.

### What changed

- `Spec2Pep._search()` now allocates three fixed-size GPU tensors:
  - `cache_tokens`: `(batch, beam, length, length)`
  - `cache_scores`: `(batch, beam, length, length)` — raw token probabilities
  - `cache_mask`: `(batch, beam, length)` — validity mask
- `_cache_finished_beams()` writes finished beams into these tensors with masked assignments instead of maintaining per-spectrum `heapq` priority queues.
- `_get_top_peptide()` flattens the cache, computes peptide scores in vectorized form, and selects the best candidate per spectrum.
- `_get_topk_beams()` uses integer division/modulo instead of `torch.unravel_index` to avoid GPU→CPU transfer.

### Hardware

All runs used a single GPU per species:

- GPU: NVIDIA GeForce RTX 4090
- CUDA: 12.4
- PyTorch: 2.6.0+cu124
- predict_batch_size: 400

### Data

Validation was performed on the **DeepNovo nine-species benchmark data**. For this PR, full test sets of the following species were used to measure speed and correctness:

- `mouse` (full test set, beam=5)
- `ricebean` (full test set, beam=5)

### Performance

| Dataset | Baseline (spectra/s) | Optimized (spectra/s) | Speedup |
|---|---|---|---|
| mouse | 88.6 | 102.2 | +15.3% |
| ricebean | 77.4 | 86.7 | +12.0% |

### Correctness

- **mouse** full beam5: all 36,907 predicted sequences identical between baseline and optimized.
- **ricebean** full beam5: 37,455 / 37,456 sequences identical. The single difference (PSM_ID 19093) is a tie between S and G resolved by floating-point order; score difference ~2.2e-8.

### Attachments

The following result files are attached to this PR:

- `mouse_beam5_full_baseline.{mztab,log}`
- `mouse_beam5_full_optimized.{mztab,log}`
- `ricebean_beam5_full_baseline.{mztab,log}`
- `ricebean_beam5_full_optimized.{mztab,log}`

## Summary by CodeRabbit

* **Performance**
  * Improved peptide prediction efficiency during beam-search decoding.
  * Accelerated selection and ranking of top candidate peptides on the GPU.

* **Bug Fixes**
  * Preserved correct stop-token handling, peptide ordering, scoring, and top-match limits.
  * Maintained reliable behavior for duplicate scores and empty results.

* **Documentation**
  * Updated the changelog with the beam-search performance improvements.

### CODE REVIEW

**Comment by sword-nan** [comment]
[baseline_full_results.zip](https://github.com/user-attachments/files/31100594/baseline_full_results.zip)
[optimized_full_results.zip](https://github.com/user-attachments/files/31100595/optimized_full_results.zip)

**Comment by bittremieux** [comment]
Hi Peng @sword-nan, thanks for the contribution! The speed-up looks nice and the functional equivalence is essential. Can you please address the CodeRabbit comments and make sure the unit tests pass before I do a manual review? Thanks.

**Comment by sword-nan** [comment]
> Hi Peng @sword-nan, thanks for the contribution! The speed-up looks nice and the functional equivalence is essential. Can you please address the CodeRabbit comments and make sure the unit tests pass before I do a manual review? Thanks.

Hi, @bittremieux 

I would like to confirm a few things.

1. For the unit tests, do I need to change the interface content of the existing unit tests? Because I changed the function interfaces.
- `_cache_finished_beams()`
- `_get_top_peptide()`
- `_get_topk_beams()`

2. Is removing duplicate peptides necessary? In auto-regressive beam-search implementation, each hypothesis is generated by appending one token to a unique prefix, so identical sequences should not naturally arise within the same beam. The duplicate check in the previous implementation was therefore mainly a safeguard to ensure that the final `top_match` predictions were distinct. If the deduplication is indeed necessary, I will incorporate it into the current implementation.

**Comment by bittremieux** [comment]
> For the unit tests, do I need to change the interface content of the existing unit tests?

Yes, please make the unit tests match the implemented version.

> Is removing duplicate peptides necessary?

You're right that this was mainly a sanity check, and empirically this shouldn't occur. So I don't think it's necessary, just close the comment with a brief justification so that we have a record of the reasoning if that's ever needed in the future.

Thanks!

File: casanovo/denovo/model.py [hunk 1]
```diff
@@ -481,102 +501,74 @@ def _cache_finished_beams(
...
+        cache_scores[:, :, step, : step + 1] = torch.where(
+            write_mask,
+            raw_scores,
+            cache_scores[:, :, step, : step + 1],
+        )
+        cache_mask[:, :, step] = torch.where(
+            beams_to_cache.view(batch, beam),
+            torch.ones_like(cache_mask[:, :, step]),
+            cache_mask[:, :, step],
+        )
```
[the remark this answers is not in the document]
Comment (sword-nan) [reply casanovo/denovo/model.py]: Thanks for the suggestion. I consider this a performance optimization rather than a correctness fix. The current implementation is correct and keeps the code straightforward, so we prefer to leave it as-is for now. We can revisit if profiling shows this step is a bottleneck.

File: casanovo/denovo/model.py [hunk 2]
```diff
@@ -691,52 +682,125 @@ def _get_topk_beams(
...
-                    for pep_score, _, aa_scores, pred_tokens in heapq.nlargest(
-                        self.top_match, peptides
-                    )
-                ]
-            else:
-                yield []
+                )
+
+                if len(pred_peptides) >= self.top_match:
+                    break
```
[the remark this answers is not in the document]
Comment (sword-nan) [reply casanovo/denovo/model.py:779]: As discussed with @bittremieux, duplicate removal is not necessary here. In autoregressive beam search, identical full sequences do not naturally arise within the same beam because each hypothesis extends a unique prefix. The previous duplicate check was mainly a safeguard, and I am intentionally dropping it in the vectorized implementation.

File: tests/unit_tests/test_unit.py [hunk 3]
```diff
@@ -2702,13 +2707,19 @@ def test_cache_finished_beams(tiny_config):
...
+        batch,
+        beam,
     )
 
-    cached = [
-        pep for (_, _, _, pep) in pred_cache[0] if torch.equal(pep, true_tok)
-    ]
-    assert len(cached) == 1
+    assert cache_mask[0, 1, step]
+    assert torch.equal(cache_tokens[0, 1, step, : step + 1], true_tok)
```
[the remark this answers is not in the document]
Comment (sword-nan) [reply tests/unit_tests/test_unit.py]: Thanks for the suggestion. The current test verifies the primary write path for the selected beam. The masked-write and score-gather behavior are correct in the implementation, and we prefer to keep the test minimal unless the maintainers want stricter contracts.

File: tests/unit_tests/test_unit.py [hunk 4]
```diff
@@ -2719,20 +2730,35 @@ def test_get_top_peptide_ranking(tiny_config):
...
-        cache[0], (0.94, 0.3, [0.94] * 4, model.tokenizer.tokenize("PEPP")[0])
+    cache_mask = torch.zeros(
+        (batch, beam, length), dtype=torch.bool, device=device
     )
 
-    result = next(model._get_top_peptide(cache))
-    assert result[0][-1] == "PEPK"
+    seqs = ["PEPY", "PEPK", "PEPP"]
+    raw_probs = [0.93, 0.95, 0.94]
+    for b, (seq, raw) in enumerate(zip(seqs, raw_probs)):
```
[the remark this answers is not in the document]
Comment (sword-nan) [reply tests/unit_tests/test_unit.py:2743]: Thanks for the suggestion. The fixture lengths are guaranteed to match in this test, and the existing test suite consistently uses plain zip() without strict=True. We prefer to leave it as-is to keep the change minimal and maintain stylistic consistency.

**Comment by sword-nan** [comment]
The reported speeds (e.g., ~87 spectra/s for ricebean and ~101 spectra/s for mouse) are computed from Casanovo’s own Time Elapsed measurement, i.e., number of successfully predicted spectra / Time Elapsed. This excludes spectra that Casanovo skips as invalid.

File: casanovo/denovo/model.py [hunk 5]
```diff
@@ -481,102 +502,74 @@ def _cache_finished_beams(
...
-            is stored.
+        cache_tokens : torch.Tensor of shape
+            (n_spectra, n_beams, max_length, max_length)
+            Tensor cache for predicted tokens of finished beams.
+        cache_scores : torch.Tensor of shape
+            (n_spectra, n_beams, max_length, max_length)
+            Tensor cache for raw token probabilities of finished beams.
+        cache_mask : torch.Tensor of shape (n_spectra, n_beams, max_length)
+            Boolean mask indicating valid cache entries.
+        batch : int
```
Comment (bittremieux) [inline casanovo/denovo/model.py]: **question:** Can't these two values be derived from the cache tensor dimensions?

File: casanovo/denovo/model.py [hunk 6]
```diff
@@ -481,102 +502,74 @@ def _cache_finished_beams(
...
-                # maximum the number of beams elements).
-                if len(pred_cache[spec_idx]) < self.n_beams:
-                    heapadd = heapq.heappush
-                else:
-                    heapadd = heapq.heappushpop
-
-                heapadd(pred_cache[spec_idx], peptide_entry)
+        # Softmax over the vocabulary and gather the probability of each
+        # selected token in one shot.
+        smx = torch.softmax(
```
Comment (bittremieux) [inline casanovo/denovo/model.py]: **nitpick:** Use `self-softmax` as before (more robust in case that calculation should change in the future).

**Review by bittremieux — CHANGES_REQUESTED** [review]
This looks very good, I just have two minor comments.

File: casanovo/denovo/model.py (hunk 5 above)
Comment (sword-nan) [reply casanovo/denovo/model.py]: I’ve simplified the cache interface:

- Removed batch and beam from _cache_finished_beams; they are now derived directly from cache_tokens.shape.
- Removed cache_mask from both _cache_finished_beams and _get_top_peptide; the validity mask is now derived inside _get_top_peptide as cache_scores.any(dim=-1). This is safe because cached values are softmax probabilities (strictly positive for valid slots, zero for unwritten slots).

The call sites and unit tests have been updated accordingly. All relevant tests pass.

File: casanovo/denovo/model.py (hunk 6 above)
Comment (sword-nan) [reply casanovo/denovo/model.py]: I now use the model’s configured self.softmax instead of the functional torch.softmax. Since self.softmax operates on dim 2, I permute the vocabulary dimension to axis 2, apply the module, and permute back:
```python
scores_view = scores[:, : step + 1, :].view(batch, beam, step + 1, vocab)
smx = self.softmax(scores_view.transpose(2, 3)).transpose(2, 3)
```

**Review by bittremieux — APPROVED** [review]
Looks good. Thanks a lot for the contribution!
